File: plugins/collector/event/apps/web/nginx.go
package web

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Nginx struct {
	version    string
	reg        *regexp.Regexp
	regTengine *regexp.Regexp
}

func (Nginx) Name() string { return "nginx" }

func (Nginx) Type() string { return "web" }

func (n Nginx) Version() string { return n.version }

func (Nginx) Match(p *process.Process) bool {
	// As default, nginx runs with master_process
	// ignore the other processes, and report the worker process if we need
	return p.Name == "nginx" && strings.Contains(p.Argv, "master")
}

// Tengine matches too
func (n *Nginx) Run(p *process.Process) (m map[string]string, err error) {
	result, err := apps.Execute(p, "-v")
	if n.reg == nil {
		n.reg = regexp.MustCompile(`nginx\/(\d+\.)+\d+`)
		n.regTengine = regexp.MustCompile(`Tengine\/(\d+\.)+\d+`)
	}
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	if s := n.regTengine.FindString(result); s != "" {
		err = apps.ErrIgnore
		return
	}
	n.version = strings.TrimPrefix(str, "nginx/")
	m = make(map[string]string)
	if strings.Contains(p.Argv, "master process") {
		m["process_type"] = "master"
	} else {
		m["process_type"] = "worker"
	}
	return
}

func init() { apps.Regist(&Nginx{}) }

File: plugins/collector/src/event/apps/web/nginx.rs
use crate::event::apps::{execute, find_version, AppProc, IApp, ERR_IGNORE};
use std::collections::HashMap;

pub struct Nginx {
    version: String,
}

impl Nginx {
    pub fn new() -> Self { Nginx { version: String::new() } }
}

impl IApp for Nginx {
    fn name(&self)     -> &'static str { "nginx" }
    fn app_type(&self) -> &'static str { "web" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "nginx" && p.argv.contains("master")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        // Ignore if the binary is actually Tengine
        if out.contains("Tengine/") {
            return Err(anyhow::anyhow!(ERR_IGNORE));
        }
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        let mut m = HashMap::new();
        m.insert(
            "process_type".into(),
            if p.argv.contains("master process") { "master" } else { "worker" }.into(),
        );
        Ok(m)
    }
}
